### PR TITLE
Email notifications events 

### LINKED ISSUE #911: Email notifications events
⚠️ After #910 only

### Project ✅

- parent/children was changed: owner, watcher
- edited: watcher, parent owner (recursive)
- transfered: new owner

### Goal ✅

- created: project owner (recursive), project watcher, goal owner, participant  
- state changed: owner, issuer, participant, watcher, project owner (recursive) 
- edited: nothing for author, owner, issuer, participant, watcher, project owner (recursive) 
- deleted(archived): nothing for author, owner, issuer, participant, watcher, project owner (recursive) 

NB: not same templates in a row

### Comment ✅ 

- created: goal owner, goal issuer, goal participant, goal watcher (nb: comment author is goal watcher automatically — check), project owner, project watcher

### User

- invited
- goal assigned/unassigned ✅ 
- you are participant/not participant

Templates: https://github.com/taskany-inc/issues/issues/912

### LINKED ISSUE #1358: Support goal estimate diff in email notification
Related #911

### PR DESCRIPTION
**Pull request by awinogradov**
## PR includes

- [x] Bug Fix
- [x] Refactor
- [x] Feature

## Related issues

Resolve #911
Resolve #1358

## QA Instructions, Screenshots, Recordings
  
image

image

image

image

### CODE REVIEW

File: src/utils/worker/resolve.ts [hunk 1]
```diff
@@ -2,4 +2,7 @@ import * as emailTemplates from './mail/templates';
 import { sendMail } from './mail';
 import { JobDataMap } from './create';
 
-export const email = ({ template, data }: JobDataMap['email']) => sendMail(emailTemplates[template](data));
+export const email = async ({ template, data }: JobDataMap['email']) => {
+    const renderedTemplate = await emailTemplates[template](data);
```
Comment (awinogradov) [inline src/utils/worker/resolve.ts:6]: I think I can rollback it and remove async from template functions

File: src/utils/worker/mail/templates.ts [hunk 2]
```diff
@@ -1,23 +1,322 @@
...
+        .split('\n')
+        .map((part: string) => `> ${part}`)
+        .join('\n');
+const renderNotice = () =>
+    '_NB: you got this email because your are owner/issuer/participant/watcher of this goal or project._';
+const renderFooter = () => `
+____
 
-interface NewCommentEmailProps {
+© 2023 Taskany inc.
```
Comment (9teen90nine) [inline src/utils/worker/mail/templates.ts:23]: this probably should be a variable 😄

File: src/utils/worker/mail/templates.ts [hunk 3]
```diff
@@ -1,23 +1,322 @@
...
+interface GoalUnassignedEmailProps {
+    to: SendMailProps['to'];
+    shortId: string;
+    title: string;
+    author?: string;
+}
+
+// TODO: send notification to issuer if he is not author of changes
+export const goalUnassigned = async ({ to, shortId, title, author = 'Somebody' }: GoalUnassignedEmailProps) => {
+    const subject = `ℹ️ You was unassigned from #${shortId}`;
```
Comment (9teen90nine) [inline src/utils/worker/mail/templates.ts:273]: you were

File: src/utils/worker/mail/templates.ts [hunk 4]
```diff
@@ -1,23 +1,322 @@
...
+interface GoalAssignedEmailProps {
+    to: SendMailProps['to'];
+    shortId: string;
+    title: string;
+    author?: string;
+}
+
+// TODO: send notification to issuer if he is not author of changes
+export const goalAssigned = async ({ to, shortId, title, author = 'Somebody' }: GoalAssignedEmailProps) => {
+    const subject = `ℹ️ You was assigned to #${shortId}`;
```
Comment (9teen90nine) [inline src/utils/worker/mail/templates.ts:248]: you were

File: src/utils/worker/mail/templates.ts [hunk 5]
```diff
@@ -1,23 +1,322 @@
...
+const withBaseTmplStyles = (html: string) =>
+    `${html} <style>blockquote { padding: 5px 10px; margin: 0 0 20px; border-left: 5px solid #eee }</style>`;
 const absUrl = (s: string) => `${process.env.NEXTAUTH_URL}${s}`;
+const renderQuote = (quote: string) =>
+    quote
+        .split('\n')
+        .map((part: string) => `> ${part}`)
+        .join('\n');
+const renderNotice = () =>
+    '_NB: you got this email because your are owner/issuer/participant/watcher of this goal or project._';
```
Comment (9teen90nine) [inline src/utils/worker/mail/templates.ts:19]: your owner -> you're the owner

File: src/utils/worker/mail/templates.ts [hunk 6]
```diff
@@ -1,23 +1,322 @@
...
+    author?: string;
+}
+
+// TODO: send notification to issuer if he is not author of changes
+export const goalAssigned = async ({ to, shortId, title, author = 'Somebody' }: GoalAssignedEmailProps) => {
+    const subject = `ℹ️ You was assigned to #${shortId}`;
+    const html = md.render(`
+🧑‍💻 **${author}** assigned goal **[${shortId}: ${title}](${absUrl(
+        `/goals/${shortId}`,
+    )})** on you. Congrats and good luck! 🎉
```
Comment (9teen90nine) [inline src/utils/worker/mail/templates.ts:252]: to you

**Review by 9teen90nine — COMMENTED** [review]
some mild corrections

File: src/utils/worker/resolve.ts (hunk 1 above)
Comment (9teen90nine) [reply src/utils/worker/resolve.ts:6]: It probably doesn't matter since it's a background job anyway

**Review by 9teen90nine — APPROVED** [review]

**Review by LamaEats — APPROVED** [review]

**Pull request merged**